README.md
# Encode Email plugin for Craft CMS 3.x

Protect email addresses in your templates from robots.

This is a quick port of Barrel Strength's [Encode Email](https://sprout.barrelstrengthdesign.com/craft-plugins/encode-email) for Craft 2.x. *This repository will likely be deleted once that version is updated for Craft 3.*


## Requirements

This plugin requires Craft CMS 3.0.0-beta.23 or later.

## Installation

The only way to install this plugin is to add this repository as a dependancy to your project. I don't want to submit this to Packagist since I plan on removing once Barrel Strength publishes their update. *Keep in mind this plugin will not be around long!*

1. Open your terminal and go to your Craft project:

        cd /path/to/project

2. Open your `composer.json` file and locate the `repositories` section. Add the following:

        {
        	"type": "vcs",
        	"url": "https://github.com/mikestecker/craft3-encodeemail"
        }

3. Then tell Composer to load the plugin:

        composer require mikestecker/craft3-encodeemail:dev-master

4. In the Control Panel, go to Settings → Plugins and click the “Install” button for Encode Email.

## Encode Email Overview

Encode Email makes multiple email obfuscation filters available in your templates. Currently supported are:

1. encode
2. rot13
3. entities

Note: All filters return raw HTML so make sure you know your data is clean before you use these filters.

## Using Encode Email

### Encode Filter

The `encode` filter will encode your email address and make it harder to be picked up off the page by spambots. Under the hood, the `encode` filter uses the same encoding technique as the Rot13 filter below. `encode` is just an easier name to remember.

### Rot13

The `rot13` filter encodes the string you pass to it using the [Rot13](https://en.wikipedia.org/wiki/ROT13) cipher and returns a javascript tag to decode it on the page.

To use it, you must first create a variable with the entire string you want to encode as a Twig string, and then run the `rot13` filter on the variable you created:

```
{% set email = "<a href='mailto:you@example.com'>Your Name</a>" %}

{{ email | rot13 }}
```

To use this filter for email, your entire link tag must be wrapped in the filter. This is necessary because the filter returns a script tag and will break your link tag if you only include the email address itself. The `rot13` filter will return code similar to the following example in your page code, however, your visitors will see the link tag as you defined it above.

```
<script type="text/javascript">document.write("<n uers=‘znvygb:lbh@rknzcyr.pbz’>Lbhe Anzr</n>".replace(/[a-zA-Z]/g, function(c){return String.fromCharCode((c<="Z"?90:122)>=(c=c.charCodeAt(0)+13)?c:c-26);}));</script>
```

### HTML Entities

The entities filter encodes your content into HTML Entities. If you filter your email information with it like so:

```
<a href='{{ "mailto:you@example.com" | entities }}'>Your Name</a>
```

The code on your page will be output as HTML Entities, but your visitors will see the information that you defined above:

```
<a href='&#109;&#97;&#105;&#108;&#116;&#111;&#58;&#121;&#111;&#117;&#64;&#101;&#120;&#97;&#109;&#112;&#108;&#101;&#46;&#99;&#111;&#109;'>Your Name</a>
```

## Encode Email Roadmap

Some things to do, and ideas for potential features:

* Nothing.

Brought to you by [Mike Stecker](http://github.com/mikestecker)


CHANGELOG.md
# Encode Email Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) and this project adheres to [Semantic Versioning](http://semver.org/).

## 1.0.0 - 2017-09-11
### Added
- Initial release


src/EncodeEmail.php
<?php
/**
 * Encode Email plugin for Craft CMS 3.x
 *
 * Protect email addresses in your templates from robots.
 *
 * @link      http://github.com/mikestecker
 * @copyright Copyright (c) 2017 Mike Stecker
 */

namespace mikestecker\encodeemail;

use mikestecker\encodeemail\twigextensions\EncodeEmailTwigExtension;

use Craft;
use craft\base\Plugin;
use craft\services\Plugins;
use craft\events\PluginEvent;

use yii\base\Event;

/**
 * Craft plugins are very much like little applications in and of themselves. We’ve made
 * it as simple as we can, but the training wheels are off. A little prior knowledge is
 * going to be required to write a plugin.
 *
 * For the purposes of the plugin docs, we’re going to assume that you know PHP and SQL,
 * as well as some semi-advanced concepts like object-oriented programming and PHP namespaces.
 *
 * https://craftcms.com/docs/plugins/introduction
 *
 * @author    Mike Stecker
 * @package   EncodeEmail
 * @since     1.0.0
 *
 */
class EncodeEmail extends Plugin
{
    // Static Properties
    // =========================================================================

    /**
     * Static property that is an instance of this plugin class so that it can be accessed via
     * EncodeEmail::$plugin
     *
     * @var EncodeEmail
     */
    public static $plugin;

    // Public Methods
    // =========================================================================

    /**
     * Set our $plugin static property to this class so that it can be accessed via
     * EncodeEmail::$plugin
     *
     * Called after the plugin class is instantiated; do any one-time initialization
     * here such as hooks and events.
     *
     * If you have a '/vendor/autoload.php' file, it will be loaded for you automatically;
     * you do not need to load it in your init() method.
     *
     */
    public function init()
    {
        parent::init();
        self::$plugin = $this;

        // Add in our Twig extensions
        Craft::$app->view->twig->addExtension(new EncodeEmailTwigExtension());

        // Do something after we're installed
        Event::on(
            Plugins::class,
            Plugins::EVENT_AFTER_INSTALL_PLUGIN,
            function (PluginEvent $event) {
                if ($event->plugin === $this) {
                    // We were just installed
                }
            }
        );

/**
 * Logging in Craft involves using one of the following methods:
 *
 * Craft::trace(): record a message to trace how a piece of code runs. This is mainly for development use.
 * Craft::info(): record a message that conveys some useful information.
 * Craft::warning(): record a warning message that indicates something unexpected has happened.
 * Craft::error(): record a fatal error that should be investigated as soon as possible.
 *
 * Unless `devMode` is on, only Craft::warning() & Craft::error() will log to `craft/storage/logs/web.log`
 *
 * It's recommended that you pass in the magic constant `__METHOD__` as the second parameter, which sets
 * the category to the method (prefixed with the fully qualified class name) where the constant appears.
 *
 * To enable the Yii debug toolbar, go to your user account in the AdminCP and check the
 * [] Show the debug toolbar on the front end & [] Show the debug toolbar on the Control Panel
 *
 * http://www.yiiframework.com/doc-2.0/guide-runtime-logging.html
 */
        Craft::info(
            Craft::t(
                'encode-email',
                '{name} plugin loaded',
                ['name' => $this->name]
            ),
            __METHOD__
        );
    }

    // Protected Methods
    // =========================================================================

}


src/twigextensions/EncodeEmailTwigExtension.php
<?php
/**
 * Encode Email plugin for Craft CMS 3.x
 *
 * Protect email addresses in your templates from robots.
 *
 * @link      http://github.com/mikestecker
 * @copyright Copyright (c) 2017 Mike Stecker
 */

namespace mikestecker\encodeemail\twigextensions;

use mikestecker\encodeemail\EncodeEmail;

use Craft;

/**
 * Twig can be extended in many ways; you can add extra tags, filters, tests, operators,
 * global variables, and functions. You can even extend the parser itself with
 * node visitors.
 *
 * http://twig.sensiolabs.org/doc/advanced.html
 *
 * @author    Mike Stecker
 * @package   EncodeEmail
 * @since     1.0.0
 */
class EncodeEmailTwigExtension extends \Twig_Extension
{

    private $count = 1;


    public function getName()
    {
        return 'EncodeEmail';
    }

    /**
     * Returns an array of Twig filters, used in Twig templates via:
     *
     *      {{ 'something' | someFilter }}
     *
     * @return array
     */
    public function getFilters()
    {
        return array(
          new \Twig_SimpleFilter('encode', [$this, 'encode'], array('is_safe' => array('html'))),
          new \Twig_SimpleFilter('rot13', [$this, 'rot13'], array('is_safe' => array('html'))),
          new \Twig_SimpleFilter('entities', [$this, 'entities'], array('is_safe' => array('html'))),
        );
    }

    /**
     * Encode string using Rot13.  This function does the same as the Rot13 function
     * and only exists as it may be easier for some folks to remember 'encode'
     *
     * @param  string $string Value to be encoded
     * @return mixed          Returns Rot13 encoded string
     */
    public function encode($string)
    {

      return $this->_encodeStringRot13($string);
    }

    /**
     * Encode string using Rot13.
     *
     * @param  string $string Value to be encoded
     * @return mixed          Returns Rot13 encoded string
     */
    public function rot13($string)
    {
      return $this->_encodeStringRot13($string);
    }

    /**
     * Encode string using HTML Entities.
     *
     * @param  string $string Value to be encoded
     * @return mixed          Returns string encoded as HTML Entities
     */
    public function entities($string)
    {
      return $this->_encodeHtmlEntities($string);
    }

    /**
     * Returns a rot13 encrypted string as well as a JavaScript decoder function.
     * http://snipplr.com/view/6037/
     *
     * @param  string $string Value to be encoded
     * @return mixed          An encoded string and javascript decoder function
     */
    private function _encodeStringRot13($string)
    {;
      $rot13encryptedString = str_replace('"', '\"', str_rot13($string));

      $uniqueId = uniqid();
      $countId  = $this->count++;
      $ajaxId   = (Craft::$app->getRequest()->getIsAjax()) ? '-ajax' : '';

      $encodeId = 'encodeemail-' . $uniqueId . '-' . $countId . $ajaxId;

      $encodedString = '
    <span id="' . $encodeId . '"></span>
    <script type="text/javascript">
      var encodeemailRot13String = "' . $rot13encryptedString . '";
      var encodeemailRot13 = encodeemailRot13String.replace(/[a-zA-Z]/g, function(c){return String.fromCharCode((c<="Z"?90:122)>=(c=c.charCodeAt(0)+13)?c:c-26);});
      document.getElementById("' . $encodeId . '").innerHTML =
      encodeemailRot13;
        </script>';

      return $encodedString;
    }

    /**
     * Returns a string converted to html entities
     * http://goo.gl/LPhtJ
     *
     * @param  string $string Value to be encoded
     * @return mixed          Returns a string converted to html entities
     */
    private function _encodeHtmlEntities($string)
    {
      $string = mb_convert_encoding($string, 'UTF-32', 'UTF-8');
      $t      = unpack("N*", $string);
      $t      = array_map(function ($n)
      {
        return "&#$n;";
      }, $t);

      return implode("", $t);
    }
}
